source/fastq/fastq.d

```diff
@@ -0,0 +1,322 @@
+private import std.algorithm : find, splitter, filter;
+private import std.conv: to;
+private import std.ascii: isWhite;
+private import std.string : stripLeft, stripRight, strip;
+private import std.experimental.logger;
+import fastaq.common.utils;
+
+struct FastqToken
+{
+  size_t endPos;
+  BufRef entryid;
+  BufRef[] hfields;
+  BufRef sequence;
+  BufRef[] extras;
+  BufRef optional;
+  BufRef quality;
+
+  void release(size_t elements)
+  {
+    endPos -= elements;
+    entryid.release(elements);
+    sequence.release(elements);
+    optional.release(elements);
+    quality.release(elements);
+    foreach(ref f; hfields) f.release(elements);
+    foreach(ref f; extras) f.release(elements);
+  }
+
+  auto value(B)(B buf)
+  {
+    FastqConcreteToken!B result;
+    result.entryid = entryid.value(buf);
+    result.hfields = new B[hfields.length];
+    result.extras = new B[extras.length];
+    foreach(i, ref f; hfields)
+      result.hfields[i] = f.value(buf);
+    foreach(j, ref e; extras)
+      result.extras[j] = e.value(buf);
+    result.sequence = sequence.value(buf);
+    result.optional = optional.value(buf);
+    result.quality = quality.value(buf);
+    return result;
+  }
+}
+
+struct FastqConcreteToken(R)
+{
+  R entryid;
+  R[] hfields;
+  R[] extras;
+  R sequence;
+  R optional;
+  R quality;
+
+}
+
+auto tokenParser(Chain,
+                 char fieldsep = ' ',
+                 char subfs = ':')(Chain c) if (isIopipe!Chain && isSomeChar!(ElementEncodingType!(WindowType!Chain)))
+  {
+    auto lines = c.byLine;
+    alias ChainType = typeof(lines);
+    size_t lineCount = 0;
+    static struct Result
+    {
+      ChainType chain;
+      size_t pos;
+      alias chain this;
+
+      FastqToken nextToken()
+      {
+        // pos must start with a start identifier
+        if(pos == chain.window.length)
+          return FastqToken.init;
+        logf("Pos: %s, window length: %s, window: %s", pos, chain.window.length, chain.window );
+        assert(chain.window[pos] == '@', "Got this: " ~ chain.window[pos]);
+        // the header is the current line
+        pos += 1; // jump over the '@'
+        FastqToken result;
+        // parse header line
+        auto lazyArr = chain.window[pos .. $].stripRight.splitter(fieldsep);
+        if(!lazyArr.empty)
+          {
+            auto firstElem = lazyArr.front;
+            result.entryid = BufRef(pos, firstElem.length);
+            auto i = 0;
+            foreach(elem; lazyArr)
+              {
+                auto subFields = elem.splitter(subfs);
+                foreach(f; subFields)
+                  {
+                    if (f.empty) continue;
+                    if(i == 0)
+                      {
+                        result.hfields ~= BufRef(pos, f.length);
+                      }
+                    else
+                      {
+                        result.extras ~= BufRef(pos, f.length);
+                      }
+                    pos += f.length + 1;
```

Comment:
This is so ugly, I hate the way this works (same in fasta). I want to create a new mechanism to parse out BufRefs, I'll work on that, and then you can use it in both places for much better parsing.

Reply:
I'm going to stop reviewing here, because with a new mechanism for parsing, this code is going to look a lot different (better).

Reply:
Looking forward to the new mechanism. I don't not have to use this for my work right now so I can wait a bit.